base/linalg/dense.jl

```diff
@@ -718,14 +718,16 @@ function cos(A::AbstractMatrix{<:Real})
     if issymmetric(A)
         return full(cos(Symmetric(A)))
     end
-    return real(exp!(im*A))
+    T = complex(float(eltype(A)))
+    return real(exp!(convert(AbstractArray{T}, im*A)))
 end
 function cos(A::AbstractMatrix{<:Complex})
     if ishermitian(A)
         return full(cos(Hermitian(A)))
     end
-    X = exp!(im*A)
-    X .= (X .+ exp!(-im*A)) ./ 2
+    T = complex(float(eltype(A)))
+    X = exp!(convert(AbstractArray{T}, im*A))
+    X .= (X .+ exp!(convert(AbstractArray{T}, -im*A))) ./ 2
```

Comment:
See above.  `T.(im .* A)` and `convert.(T, .- im .* A)`, maybe?  Or
```
X = exp!(T.(im .* A))
@. X = (X + $exp!(T(-im*A))) / 2
```
if you want to reduce the number of dots in the second line